stan/math/prim/fun/unit_vector_constrain.hpp

```diff
@@ -17,30 +17,27 @@ namespace math {
  * href="https://en.wikipedia.org/wiki/N-sphere#Generating_random_points">the
  * Wikipedia page on generating random points on an N-sphere</a>.
  *
- * @tparam EigMat type inheriting from `EigenBase` that does not have an fvar
+ * @tparam T type inheriting from `EigenBase` that does not have an fvar
  *  scalar type.
  * @param y vector of K unrestricted variables
  * @return Unit length vector of dimension K
  */
 template <typename T, require_eigen_col_vector_t<T>* = nullptr,
           require_not_vt_autodiff<T>* = nullptr>
-inline auto unit_vector_constrain(const T& y) {
+inline plain_type_t<T> unit_vector_constrain(const T& y) {
   using std::sqrt;
   check_nonzero_size("unit_vector_constrain", "y", y);
-  return make_holder(
-      [](const auto& y_ref) {
-        value_type_t<T> SN = dot_self(y_ref);
-        check_positive_finite("unit_vector_constrain", "norm", SN);
-        return y_ref / sqrt(SN);
-      },
-      to_ref(y));
+  auto&& y_ref = to_ref(y);
+  value_type_t<T> SN = dot_self(y_ref);
+  check_positive_finite("unit_vector_constrain", "norm", SN);
+  return y_ref.array() / sqrt(SN);
```

Comment:
[comment]
[comment]
Ack, we don't have a Euclidean length function built-in, but we should.  I created an issue for it: #2562

```
sqrt(dot_self(y)) == length(y)
```